src/frontend/src/optimizer/rule/distinct_agg.rs

```diff
@@ -66,37 +55,68 @@ impl DistinctAggRule {
+    /// `Expand` if there is only one `subset`.
     fn build_expand(
         input: PlanRef,
-        group_keys: &[usize],
-        distinct_aggs: &[PlanAggCall],
-        non_distinct_aggs: &[PlanAggCall],
-    ) -> PlanRef {
+        group_keys: &mut Vec<usize>,
+        agg_calls: &mut Vec<PlanAggCall>,
+    ) -> Option<(PlanRef, Vec<usize>, bool)> {
+        let input_schema_len = input.schema().len();
         // each `subset` in `column_subsets` consists of `group_keys`, `agg_call`'s input indices
         // and the input indices of `agg_call`'s `filter`.
         let mut column_subsets = vec![];
+        // flag values of distinct aggregates.
+        let mut flag_values = vec![];
+        // mapping from `subset` to `flag_value`, which is used to deduplicate `column_subsets`.
+        let mut hash_map = HashMap::new();
+        let (distinct_aggs, non_distinct_aggs): (Vec<_>, Vec<_>) =
+            agg_calls.iter().partition(|agg_call| agg_call.distinct);
+        if distinct_aggs.is_empty() {
+            return None;
+        }
 
         if !non_distinct_aggs.is_empty() {
-            column_subsets.push({
-                let mut subset = FixedBitSet::from_iter(group_keys.to_owned());
+            let subset = {
+                let mut subset = FixedBitSet::from_iter(group_keys.iter().cloned());
                 non_distinct_aggs.iter().for_each(|agg_call| {
                     subset.extend(agg_call.input_indices());
                 });
                 subset.ones().collect_vec()
-            });
+            };
+            hash_map.insert(subset.clone(), 0);
+            column_subsets.push(subset);
         }
 
         distinct_aggs.iter().for_each(|agg_call| {
-            column_subsets.push({
-                let mut subset = FixedBitSet::from_iter(group_keys.to_owned());
+            let subset = {
+                let mut subset = FixedBitSet::from_iter(group_keys.iter().cloned());
                 subset.extend(agg_call.input_indices());
                 subset.ones().collect_vec()
-            });
+            };
+            if let Some(i) = hash_map.get(&subset) {
+                flag_values.push(*i);
+            } else {
+                let flag_value = column_subsets.len();
+                flag_values.push(flag_value);
+                hash_map.insert(subset.clone(), flag_value);
+                column_subsets.push(subset);
+            }
         });
 
-        LogicalExpand::create(input, column_subsets)
+        if column_subsets.len() == 1 {
+            // no need to have expand if there is only one distinct aggregates.
+            return Some((input, flag_values, false));
+        }
+        let expand = LogicalExpand::create(input, column_subsets);
+        // manual version of column pruning for expand.
+        let project = Self::build_project(input_schema_len, expand, group_keys, agg_calls);
+        Some((project, flag_values, true))
     }
 
+    /// Used to do column pruning for `Expand`.
```

Comment:
Could you elaborate why the general column pruning cannot work here? How is it simpler than doing the general one?

Reply:
I currently assume that the schema of Expand is `| expanded columns(i.e. some columns are set to null) | original columns of input | flag |` because it fits into previous definition and backend implementation of Expand. 
Given this assumption, the backend implementation of Expand is greatly simplied, otherwise we need something to record which column need to be outputed.

Reply:
I think column pruning for `Expand` is how to prune the input of `Expand` according to the required columns. However, I can not see any change about the input of `Expand`. Do I understand right?

Reply:
If we just prune the input of `Expand`, then `Expand` will still output some unused columns.
BTW, converting to distinct aggregates is done **after** column pruning now.

Reply:
I meant adding the explanation as comments to help reading 😂